File: src/test/scala/com/projecttemplategroup/projecttemplateartifact/security/AccessLogicAsUser.scala
package com.projecttemplategroup.projecttemplateartifact.security

import org.springframework.http.{HttpEntity, HttpMethod, HttpStatus, ResponseEntity}
import com.projecttemplategroup.projecttemplateartifact.BaseSpringTest

class AccessLogicAsUser extends BaseSpringTest {

  val userToken = "test-token-user"

  Feature("check user cookie token") {

    Scenario("success call public api as user") {
      val resp = restTemplate.exchange("/v1/ping", HttpMethod.GET, new HttpEntity[String](cookieHeader(userToken)), classOf[String])

      resp.getStatusCode shouldBe HttpStatus.OK
      resp.getBody shouldBe "public api pong"
    }

    Scenario("success call user api as user") {
      val resp: ResponseEntity[String] = restTemplate.exchange("/v1-usr/ping", HttpMethod.GET, new HttpEntity[String](cookieHeader(userToken)), classOf[String])

      resp.getStatusCode shouldBe HttpStatus.OK
      resp.getBody shouldBe "user pong"
    }

    Scenario("fail to call user api with invalid token") {
      val resp = restTemplate.exchange("/v1-usr/ping", HttpMethod.GET, new HttpEntity[String](cookieHeader("aaa")), classOf[String])

      resp.getStatusCode shouldBe HttpStatus.UNAUTHORIZED
      jsonAssert(
        """{
          |"status":401,
          |"error":"Unauthorized",
          |"message":""}""".stripMargin, resp.getBody, false)
    }

    Scenario("fail to call admin api as user") {
      val resp = restTemplate.exchange("/v1-adm/ping", HttpMethod.GET, new HttpEntity[String](cookieHeader(userToken)), classOf[String])

      resp.getStatusCode shouldBe HttpStatus.FORBIDDEN
      jsonAssert(
        """{
          |"status":403,
          |"error":"Forbidden",
          |"message":""}""".stripMargin, resp.getBody, false)
    }

  }
}

File: src/test/java/com/projecttemplategroup/projecttemplateartifact/security/AccessLogicAsUser.java
package com.projecttemplategroup.projecttemplateartifact.security;

import com.projecttemplategroup.projecttemplateartifact.BaseSpringTest;
import org.junit.jupiter.api.Test;
import org.springframework.http.*;

import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
//todo move to scala
class AccessLogicAsUser extends BaseSpringTest {

    @Test
    void success_ping_public_api_as_user() {
        ResponseEntity<String> resp = restTemplate.exchange("/v1/ping", HttpMethod.GET, new HttpEntity<String>(cookieHeader("test-token-user")), String.class);

        assertEquals(HttpStatus.OK, resp.getStatusCode());
        assertEquals("public api pong", resp.getBody());
    }

    @Test
    void success_ping_user_api_as_user() {
        ResponseEntity<String> resp = restTemplate.exchange("/v1-usr/ping", HttpMethod.GET, new HttpEntity<String>(cookieHeader("test-token-user")), String.class);

        assertEquals(HttpStatus.OK, resp.getStatusCode());
        assertEquals("user pong", resp.getBody());
    }

    @Test
    void fail_ping_user_api_with_invalid_token() {
        ResponseEntity<Map> resp = restTemplate.exchange("/v1-usr/ping", HttpMethod.GET, new HttpEntity<String>(cookieHeader("aaa")), Map.class);

        assertEquals(HttpStatus.UNAUTHORIZED, resp.getStatusCode());
        assertEquals("", resp.getBody().get("message"));
    }

    @Test
    void fail_ping_admin_api_as_user() {
        ResponseEntity<Map> resp = restTemplate.exchange("/v1-adm/ping", HttpMethod.GET, new HttpEntity<String>(cookieHeader("test-token-user")), Map.class);

        assertEquals(HttpStatus.FORBIDDEN, resp.getStatusCode());
        assertEquals("", resp.getBody().get("message"));
    }

}
